### ISSUE TITLE
Proposition to delete _get_func_result_for_tensor_or_weighted_tensor - [merged]

### ISSUE THREAD

**Issue by NicolasGensollen** [aramis-lab/leaspy#250]
In GitLab by @JulietteOrtholand on Mar 25, 2024, 15:51

_Merges fix_wt_event -> v2_

The function `_get_func_result_for_tensor_or_weighted_tensor` mask the fact that all the input of _nll are not Weighted tensors. 

It seems that the nll could be used for two purpose, and depending on the purpose, the input type is not the same:
- if it is used to compute regularisation for a variable, then it does not have an ancestor that is a DataVariable, then the input is not a weighted tensor,
- if it is used from the attachment, then the input is a weighted tensor as it as an ancestor which is a DataVariable. 

It might require further investigation...

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 25, 2024, 15:56

marked this merge request as **draft**

**Comment by NicolasGensollen** [comment]
Could parameters be `WeightedTensor` objects ? If yes, we should call \`\_cast_parameters_to_tensors\` before calling `_nll`, otherwise we should change the type for `torch.Tensor`.

**Comment by NicolasGensollen** [comment]
I don't think `nll_grad_value` is a WeightedTensor as `grad` returns a tuple of Tensor objects.

**Comment by NicolasGensollen** [comment]
Same question for the parameters of `regularization`

**Comment by NicolasGensollen** [comment]
You'd need to update the type of the return value.

Also, shouldn't this be : 

```python
return WeightedTensor(-1 * (log_survival + log_hazard), x.weight)
```

?

**Comment by NicolasGensollen** [comment]
Thanks @JulietteOrtholand ! 

I made some comments, mostly on type mismatch issues.

I'm a bit surprised that changing the return type of `nll` from a Tensor to a WeightedTensor doesn't break things. I'll take a closer look when I get a chance.

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 26, 2024, 15:05

Commented on [leaspy/variables/distributions.py line 134](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..387185d7f3933a36d09c7489d9fc68d11004407c#diff-4b6be71874d2c6953db4ae68f1640fa8R134)



changed this line in [version 3 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=964336012&start_sha=387185d7f3933a36d09c7489d9fc68d11004407c#30012bd69bbe81b60516fc5f8c88ec423c339f24_134_134)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 26, 2024, 15:05

added 1 commit

<ul><li>ba0ceab8 - changes after review</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=964336012&start_sha=387185d7f3933a36d09c7489d9fc68d11004407c)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 26, 2024, 15:07

Commented on [leaspy/variables/distributions.py line 238](https://github.com/aramis-lab/leaspy/compare/a308c9318af64c07f5366e571e5c05f3b3937d4c..38f786a753c6d25b15773700afebe68bd47ff9ca#diff-4b6be71874d2c6953db4ae68f1640fa8R238)



I changed it to 
`return nll, WeightedTensor(nll_grad_value, x.weight)`

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 26, 2024, 15:07

Commented on [leaspy/variables/distributions.py line 529](https://github.com/aramis-lab/leaspy/compare/a308c9318af64c07f5366e571e5c05f3b3937d4c..38f786a753c6d25b15773700afebe68bd47ff9ca#diff-4b6be71874d2c6953db4ae68f1640fa8R529)



I change the type of the return value. What you suggest for the output is precisely what the delete function was doing wrong, here the weights are applied n the hazard only and not on the total likelihood.

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 26, 2024, 15:07

Commented on [leaspy/variables/distributions.py line 134](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..387185d7f3933a36d09c7489d9fc68d11004407c#diff-4b6be71874d2c6953db4ae68f1640fa8R134)



Params should not be Weighted tensors as they are model parameters or hyperparameters, I change for torch.tensor.

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 26, 2024, 15:07

Commented on [leaspy/variables/distributions.py line 142](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..abc55488be64ff45c0f7d6fa00917d0c102951d7#diff-4b6be71874d2c6953db4ae68f1640fa8R142)



As above, I changed for torch.Tensor

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 26, 2024, 15:07

I did not change the return type of `nll` I only changed the return type of `_nll`, which is called inside of `nll`, to make them the same. un as `_nll` must be a private function, it should not be called outside right? So that might explain why it does not break things.

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 26, 2024, 15:07

Thanks for your review, I answered your questions and made the associated changes. I also made a hypothesis as to why such change does not break everything.

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 26, 2024, 16:56

added 1 commit

<ul><li>abc55488 - from _get_func to get_func</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=964516943&start_sha=ba0ceab8d6e16e70ac27d827869b795776671319)

**Comment by NicolasGensollen** [comment]
```suggestion:-0+0
            x: torch.Tensor,
```

**Comment by NicolasGensollen** [comment]
You can delete `_cast_parameters_to_tensors` and `_check_weighted_tensors_have_same_weights` as they are unused now.

**Comment by NicolasGensollen** [comment]
Tensor isn't parametric:

```suggestion:-0+0
            *params: torch.Tensor,
```

**Comment by NicolasGensollen** [comment]
```suggestion:-0+0
            *params: torch.Tensor,
```

**Comment by NicolasGensollen** [comment]
```suggestion:-0+0
            *params: torch.Tensor,
```

**Comment by NicolasGensollen** [comment]
nll is now a WeightedTensor, so it maybe:

```suggestion:-0+0
        (nll_grad_value,) = grad(nll.value, (x.value,), create_graph=x.requires_grad)
```

**Comment by NicolasGensollen** [comment]
```suggestion:-0+0
            *params: torch.Tensor,
```

**Comment by NicolasGensollen** [comment]
```suggestion:-0+0
        return WeightedTensor(nll, x.weight), WeightedTensor(z / scale, x.weight)
```

**Comment by NicolasGensollen** [comment]
Thanks @JulietteOrtholand !

I just made a few suggestions to improve typing

**Comment by NicolasGensollen** [comment]
Also, could you rebase on V2 in order to get the fix from @little_nemo ? It would make it easier to assess that the number of failing tests is decreasing.

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 09:48

Commented on [leaspy/variables/distributions.py line 142](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..abc55488be64ff45c0f7d6fa00917d0c102951d7#diff-4b6be71874d2c6953db4ae68f1640fa8R142)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=965225112&start_sha=abc55488be64ff45c0f7d6fa00917d0c102951d7#30012bd69bbe81b60516fc5f8c88ec423c339f24_142_142)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 09:48

Commented on [leaspy/variables/distributions.py line 142](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..abc55488be64ff45c0f7d6fa00917d0c102951d7#diff-4b6be71874d2c6953db4ae68f1640fa8R142)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=965225112&start_sha=abc55488be64ff45c0f7d6fa00917d0c102951d7#30012bd69bbe81b60516fc5f8c88ec423c339f24_142_142)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 09:48

Commented on [leaspy/variables/distributions.py line 134](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..abc55488be64ff45c0f7d6fa00917d0c102951d7#diff-4b6be71874d2c6953db4ae68f1640fa8R134)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=965225112&start_sha=abc55488be64ff45c0f7d6fa00917d0c102951d7#30012bd69bbe81b60516fc5f8c88ec423c339f24_134_134)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 09:48

Commented on [leaspy/variables/distributions.py line 143](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..abc55488be64ff45c0f7d6fa00917d0c102951d7#diff-4b6be71874d2c6953db4ae68f1640fa8R143)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=965225112&start_sha=abc55488be64ff45c0f7d6fa00917d0c102951d7#30012bd69bbe81b60516fc5f8c88ec423c339f24_143_142)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 09:48

Commented on [leaspy/variables/distributions.py line 161](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..abc55488be64ff45c0f7d6fa00917d0c102951d7#diff-4b6be71874d2c6953db4ae68f1640fa8R161)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=965225112&start_sha=abc55488be64ff45c0f7d6fa00917d0c102951d7#30012bd69bbe81b60516fc5f8c88ec423c339f24_161_161)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 09:48

Commented on [leaspy/variables/distributions.py line 160](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..abc55488be64ff45c0f7d6fa00917d0c102951d7#diff-4b6be71874d2c6953db4ae68f1640fa8L160)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=965225112&start_sha=abc55488be64ff45c0f7d6fa00917d0c102951d7#30012bd69bbe81b60516fc5f8c88ec423c339f24_152_152)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 09:48

Commented on [leaspy/variables/distributions.py line 376](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..abc55488be64ff45c0f7d6fa00917d0c102951d7#diff-4b6be71874d2c6953db4ae68f1640fa8R376)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=965225112&start_sha=abc55488be64ff45c0f7d6fa00917d0c102951d7#30012bd69bbe81b60516fc5f8c88ec423c339f24_376_376)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 09:48

added 1 commit

<ul><li>36af7403 - Nicolas&#39; suggestions</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=965225112&start_sha=abc55488be64ff45c0f7d6fa00917d0c102951d7)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 09:55

Commented on [leaspy/variables/distributions.py line 132](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..36af74037929061b64beb339bf69d860715045c6#diff-4b6be71874d2c6953db4ae68f1640fa8L132)



changed this line in [version 6 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=965233483&start_sha=36af74037929061b64beb339bf69d860715045c6#30012bd69bbe81b60516fc5f8c88ec423c339f24_115_108)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 09:55

Commented on [leaspy/variables/distributions.py line 270](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..36af74037929061b64beb339bf69d860715045c6#diff-4b6be71874d2c6953db4ae68f1640fa8L270)



changed this line in [version 6 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=965233483&start_sha=36af74037929061b64beb339bf69d860715045c6#30012bd69bbe81b60516fc5f8c88ec423c339f24_262_240)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 09:55

added 1 commit

<ul><li>8fa5fff2 - second review</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=965233483&start_sha=36af74037929061b64beb339bf69d860715045c6)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 09:59

added 8 commits

<ul><li>8fa5fff2...a308c931 - 2 commits from branch <code>v2</code></li><li>914d7ed5 - first draft</li><li>cdf7f916 - cleaner version</li><li>5ee54ca0 - changes after review</li><li>d93a7480 - from _get_func to get_func</li><li>91389183 - Nicolas&#39; suggestions</li><li>38f786a7 - second review</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/128/diffs?diff_id=965238565&start_sha=8fa5fff2efb63eed11b88a259c460d0ea236342a)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 10:00

Commented on [leaspy/variables/distributions.py line 132](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..36af74037929061b64beb339bf69d860715045c6#diff-4b6be71874d2c6953db4ae68f1640fa8L132)



I have suppressed those two functions

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 10:00

Commented on [leaspy/variables/distributions.py line 270](https://github.com/aramis-lab/leaspy/compare/27c57220d8e2d394c7d859635fcc65b8c16e9496..36af74037929061b64beb339bf69d860715045c6#diff-4b6be71874d2c6953db4ae68f1640fa8L270)



Done

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 10:00

resolved all threads

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Mar 27, 2024, 10:00

Thanks for your review, normally everything is ready. Just waiting for the CI to be sure !

**Comment by NicolasGensollen** [comment]
approved this merge request

**Comment by NicolasGensollen** [comment]
mentioned in commit 01384a7c86d04904ac47a167f08a24a04b2ab0df
